File: python/Camera.py
import numpy as np
import cv2
import threading

class Camera:
    def __init__(self, name, path, width=1280, height=720, fps=25, fourcc='MJPG'):
        self.name = name
        self.path = path
        self.width = width
        self.height = height
        self.fps = fps
        self.fourcc = cv2.VideoWriter_fourcc(*fourcc)
        self.cap = cv2.VideoCapture(path)
        self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, width)
        self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
        self.cap.set(cv2.CAP_PROP_FPS, fps)
        self.cap.set(cv2.CAP_PROP_FOURCC, self.fourcc)
        
        if not self.cap.isOpened():
            raise RuntimeError(f"无法打开摄像头 {name}")
        
        self.frame_count = 0
        self.frame = None
        self.running = False
        self.lock = threading.Lock()  # 添加锁用于线程同步
        self.frame_ready = threading.Event()  # 用于通知新帧可用
        
    def start(self):
        """启动摄像头线程"""
        self.running = True
        self.thread = threading.Thread(target=self._run, daemon=True)
        self.thread.start()
        
    def stop(self):
        """停止摄像头线程"""
        self.running = False
        self.frame_ready.set()  # 确保等待帧的线程被唤醒
        
        if hasattr(self, 'thread') and self.thread.is_alive():
            self.thread.join(timeout=2.0)
            if self.thread.is_alive():
                print(f"警告: 摄像头 {self.name} 线程未正常退出")
                # 强制终止线程作为最后手段
                # 注意：这可能导致资源未释放，但避免程序崩溃
                self.thread._stop()  # 不推荐但有时必要
        
        # 释放资源
        if self.cap.isOpened():
            self.cap.release()
        print(f"摄像头 {self.name} 已停止")
    
    def _run(self):
        """内部运行方法，不应直接调用"""
        while self.running and self.cap.isOpened():
            try:
                ret, frame = self.cap.read()
                if not ret:
                    print(f"摄像头 {self.name} 读取失败")
                    break
                
                with self.lock:  # 使用锁保护帧的写入
                    self.frame = frame.copy()  # 创建副本以避免数据竞争
                    self.frame_count += 1
                    # print(f"{self.name} : {self.frame_count}")
    
                self.frame_ready.set()  # 通知新帧可用
                
                # 在主线程中显示帧，而不是在摄像头线程中
                # 这样可以避免OpenCV的GUI问题
                
            except Exception as e:
                print(f"摄像头 {self.name} 出错: {e}")
                break
    
    def get_frame(self):
        """获取最新帧"""
        self.frame_ready.wait()  # 等待直到有帧可用
        with self.lock:  # 使用锁保护帧的读取
            if self.frame is not None:
                return self.frame.copy()  # 返回副本以避免数据竞争
            return None
    
    def get_frame_count(self):
        """获取帧计数"""
        with self.lock:  # 使用锁保护计数器的读取
            return self.frame_count

File: src/camera.cpp
#include "camera.hpp"

namespace stitching {
    Camera::Camera(string name, string path, int width, int height, int fps, string fourcc)
    {
        _name = name;
        _path = path;
        _width = width;
        _height = height;
        _fps = fps;
        if (fourcc.length() != 4)
        {
            _fourcc = VideoWriter::fourcc('M', 'J', 'P', 'G');
        }
        else
        {
            _fourcc = VideoWriter::fourcc(fourcc[0], fourcc[1], fourcc[2], fourcc[3]);
        }
        _cap = VideoCapture(_path, cv::CAP_V4L2);
        _cap.set(CAP_PROP_FRAME_WIDTH, _width);
        _cap.set(CAP_PROP_FRAME_HEIGHT, _height);
        _cap.set(CAP_PROP_FPS, _fps);
        _cap.set(CAP_PROP_FOURCC, _fourcc);
        cout << "Camera: " + _name + " open with path [" + _path + "]" << endl;
        cout << "Camera: " + _name + " open with width [" + to_string(_width) + "]" << endl;
        cout << "Camera: " + _name + " open with height [" + to_string(_height) + "]" << endl;
        cout << "Camera: " + _name + " open with fps [" + to_string(_fps) + "]" << endl;
        cout << "Camera: " + _name + " open with fourcc [" + to_string(_fourcc) + "]" << endl;
        if (_cap.isOpened() == false)
        {
            throw runtime_error("Camera: cannot open the [" + _name + "], with path [" + _path + "]");
        }
        _frame_counter = 0;
        _running_flag = false;
    }

    Camera::~Camera()
    {
        stop();
    }

    void Camera::start(void)
    {
        if (_running_flag == true)
        {
            cerr << "Camera: " + _name + " is running" << endl;
            return;
        }

        _running_flag = true;
        _thread_run = thread(&Camera::_run, this);

    }
    
    void Camera::stop(void)
    {
        if (_running_flag == false)
        {
            cerr << "Camera: " + _name + " has stopped" << endl;
        }

        _running_flag = false;
        _cv_frame_ready.notify_all();

        if ( _thread_run.joinable() )
        {
            _thread_run.join();
        }
        else
        {
            cerr << "Camera: " + _name + " running thread has joinned" << endl;
        }

        if ( _cap.isOpened() )
        {
            _cap.release();
        }
        else
        {

            cerr << "Camera: " + _name + " cap has released" << endl;
        }
    }


    void Camera::_run(void)
    {
        while (_running_flag && _cap.isOpened())
        {
            try
            {
                Mat temp_frame;
                bool ret = _cap.read(temp_frame);
                if (!ret || temp_frame.empty())
                {
                    // throw runtime_error("Camera: " + _name + " read frame failed");
                    std::cout << "Camera: " + _name + " read frame failed" << std::endl;
                    this_thread::sleep_for(chrono::milliseconds(1)); // 避免空转
                    continue;
                }
                {
                    // 加锁保护 _frame 资源
                    unique_lock<mutex> lock(_mtx_frame);
                    temp_frame.copyTo(_frame);
                    _frame_ready_flag = true;
                }
                _frame_counter++;
                // std::cout << "Camera: " + _name + " read frame " << _frame_counter << std::endl;
                // 通知等待帧的线程
                _cv_frame_ready.notify_one();

                // 如果需要控制帧率，可以这里加延时，例如 40ms 对应 25fps
                this_thread::sleep_for(std::chrono::milliseconds(1));
            }
            catch(const std::exception& e)
            {
                std::cerr << "Camera: " + _name + ", exception: " << e.what() << std::endl;
                exit(1);
            }
        }
    }


    Mat Camera::get_frame(void)
    {
        unique_lock<mutex> lock(_mtx_frame);  // 自动加锁，离开作用域自动解锁
        _cv_frame_ready.wait_for(lock,  chrono::milliseconds(100), [this]{ return  _frame_ready_flag; });
        Mat ret_frame = _frame.clone();
        _frame_ready_flag = false;
        return ret_frame;
    }

}